Work around a linking problem on 32-bit arches

nalind opened the pull request:
This should let us avoid triggering the linking error we were seeing in https://github.com/docker/docker/pull/18104, at least for now (future additions could trigger the bug, too).  We won't need this workaround once we switch to Go 1.6.

nalind committed:
Revert "prevent journald from being built on ARM"

This reverts commit 6f6f10a75f8b447637e8a89d685452871899e9c0, so that we
can apply a different workaround.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "workaround-go-libgcc" git@github.com:nalind/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

LK4D4 commented:
@nalind Thanks a lot!
@Govinda-Fichtner Would you mind to try this on ARM?

LK4D4 commented:
LGTM

crosbymichael commented:
LGTM

crosbymichael commented:
we can wait for @Govinda-Fichtner to confirm on arm before merge but code looks good

Govinda-Fichtner commented:
@crosbymichael Gonna test this PR right now.

Govinda-Fichtner commented:
@nalind I did clone your fork on my Scaleway ARM server and checkout out this branch. Unfortunately building a dynamic binary still fails.

But this time it is a different error:

```
docker run --rm -i --privileged -e BUILDFLAGS -e DOCKER_CLIENTONLY -e DOCKER_DEBUG -e DOCKER_ENGINE_GOARCH -e DOCKER_ENGINE_GOOS -e DOCKER_ENGINE_OSARCH -e DOCKER_EXPERIMENTAL -e DOCKER_FILE -e DOCKER_GRAPHDRIVER -e DOCKER_REMAP_ROOT -e DOCKER_STORAGE_OPTS -e DOCKER_USERLANDPROXY -e TESTDIRS -e TESTFLAGS -e TIMEOUT -v "/docker/bundles:/go/src/github.com/docker/docker/bundles" -t "docker-dev:workaround-go-libgcc" hack/make.sh dynbinary

---> Making bundle: dynbinary (in bundles/1.10.0-dev/dynbinary)
Created binary: bundles/1.10.0-dev/dynbinary/dockerinit-1.10.0-dev
Building: bundles/1.10.0-dev/dynbinary/docker-1.10.0-dev
# github.com/docker/docker/docker
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_l2d':
:(.text+0x879ac): multiple definition of `__floatdidf'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x8fc): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_i2d':
:(.text+0x8793c): multiple definition of `__floatsidf'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x88c): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__floatunsidf':
:(.text+0x8791c): multiple definition of `__aeabi_ui2d'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x86c): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_drsub':
:(.text+0x87698): multiple definition of `__aeabi_drsub'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x5e8): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_f2d':
:(.text+0x87960): multiple definition of `__extendsfdf2'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x8b0): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_dadd':
:(.text+0x876a4): multiple definition of `__adddf3'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x5f4): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__floatunsidf':
:(.text+0x8791c): multiple definition of `__floatunsidf'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x86c): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_dadd':
:(.text+0x876a4): multiple definition of `__aeabi_dadd'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x5f4): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__floatundidf':
:(.text+0x8799c): multiple definition of `__aeabi_ul2d'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x8ec): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__subdf3':
:(.text+0x876a0): multiple definition of `__aeabi_dsub'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x5f0): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__floatundidf':
:(.text+0x8799c): multiple definition of `__floatundidf'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x8ec): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_i2d':
:(.text+0x8793c): multiple definition of `__aeabi_i2d'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x88c): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_l2d':
:(.text+0x879ac): multiple definition of `__aeabi_l2d'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x8fc): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__aeabi_f2d':
:(.text+0x87960): multiple definition of `__aeabi_f2d'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x8b0): first defined here
/var/tmp/go-link-TMSqwV/000008.o: In function `__subdf3':
:(.text+0x876a0): multiple definition of `__subdf3'
/var/tmp/go-link-TMSqwV/000003.o::(.text+0x5f0): first defined here
collect2: error: ld returned 1 exit status
/usr/local/go/pkg/tool/linux_arm/5l: running gcc failed: unsuccessful exit status 0x100
Makefile:71: recipe for target 'dynbinary' failed
make: *** [dynbinary] Error 1
```

Tell me if you need more information or if I can help in some other way.

nalind commented:
Yikes, that's a lot more duplicate support functions than I got on x86.  The floating point emulation and software implementation of integer division is more than I can probably remove from one or the other cgo-using areas of the tree.  I think the only workaround we'll be able to get away with there is going to be external linking with -z muldefs.

nalind committed:
Use -z,muldefs on arm,x86

Assume that the linker can make sense of us passing in the -z,muldefs
option to tell it to ignore symbol-multiply-defined errors triggered by
https://github.com/golang/go/issues/9510.  We should be able to stop
doing this once we move to Go 1.6.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

nalind commented:
@Govinda-Fichtner Okay, I've got https://github.com/nalind/docker/tree/workaround-go-libgcc-2 using the external linker with the -z muldefs linker flag when GOARCH matches "arm*" or "386".  Any chance you can check if the right things happen on your builder?  If so, I can either push it into this PR's branch or open another one using that branch.

Govinda-Fichtner commented:
@nalind `workaround-go-libgcc-2` did not produce any errors for me:

```
docker run --rm -i --privileged -e BUILDFLAGS -e DOCKER_CLIENTONLY -e DOCKER_DEBUG -e DOCKER_ENGINE_GOARCH -e DOCKER_ENGINE_GOOS -e DOCKER_ENGINE_OSARCH -e DOCKER_EXPERIMENTAL -e DOCKER_FILE -e DOCKER_GRAPHDRIVER -e DOCKER_REMAP_ROOT -e DOCKER_STORAGE_OPTS -e DOCKER_USERLANDPROXY -e TESTDIRS -e TESTFLAGS -e TIMEOUT -v "/root/docker/bundles:/go/src/github.com/docker/docker/bundles" -t "docker-dev:workaround-go-libgcc-2" hack/make.sh dynbinary


---> Making bundle: dynbinary (in bundles/1.10.0-dev/dynbinary)
Created binary: bundles/1.10.0-dev/dynbinary/dockerinit-1.10.0-dev
Building: bundles/1.10.0-dev/dynbinary/docker-1.10.0-dev
Created binary: bundles/1.10.0-dev/dynbinary/docker-1.10.0-dev
```

LK4D4 commented:
@nalind @Govinda-Fichtner sounds great

calavera commented:
@nalind can you port the changes in that other branch here? otherwise, please, close this PR and open a new one for that other branch.

nalind commented:
Okay, switching over to that changeset for this PR.

nalind commented:
Trying without the -linkmode external, since I think the build environment is missing some bits which would be needed to do things using make.sh's binary/dynbinary/cross targets.

estesp commented:
@Govinda-Fichtner just to make sure--the code now in this PR is what you've tested and validated works for your ARM build?  @nalind looks complete, but just need a final OK that this PR is the final version that fixes the link issues.

tiborvass commented:
@Govinda-Fichtner let us know if this looks good to you

thaJeztah commented:
ping @Govinda-Fichtner can you have a look please?

also /cc @DieterReuter

Govinda-Fichtner commented:
@thaJeztah Yes. Gonna test it again this evening. The last test was quite some time ago.

Govinda-Fichtner commented:
@estesp The PR works for me. I can build a dynamic binary without error.

```
docker run --rm -i --privileged -e BUILDFLAGS -e DOCKER_CLIENTONLY -e DOCKER_DEBUG -e DOCKER_ENGINE_GOARCH -e DOCKER_ENGINE_GOOS -e DOCKER_ENGINE_OSARCH -e DOCKER_EXPERIMENTAL -e DOCKER_FILE -e DOCKER_GRAPHDRIVER -e DOCKER_REMAP_ROOT -e DOCKER_STORAGE_OPTS -e DOCKER_USERLANDPROXY -e TESTDIRS -e TESTFLAGS -e TIMEOUT -v "/root/docker/bundles:/go/src/github.com/docker/docker/bundles" -t "docker-dev:workaround-go-libgcc" hack/make.sh dynbinary

bundles/1.10.0-dev already exists. Removing.

---> Making bundle: dynbinary (in bundles/1.10.0-dev/dynbinary)
Created binary: bundles/1.10.0-dev/dynbinary/dockerinit-1.10.0-dev
Building: bundles/1.10.0-dev/dynbinary/docker-1.10.0-dev
Created binary: bundles/1.10.0-dev/dynbinary/docker-1.10.0-dev
```

estesp commented:
LGTM

thaJeztah commented:
Thanks @nalind, and thanks so much for testing @Govinda-Fichtner!
